carsus/io/output/base.py

```diff
@@ -240,188 +260,222 @@ def calculate_collisional_strength(row, temperatures,
+            levels = levels.drop(levels[mask].index)
+
+        levels = levels[['atomic_number', 'ion_number', 'g', 'energy', 
+                         'ds_id']]
         levels = levels.reset_index()
 
         return levels
 
 
     def _get_all_lines_data(self):
-        """ Returns the same output than `AtomData._get_all_lines_data()` """
-
-        gf = self.gfall_reader
-        ch = self.chianti_reader
-
-        start = 1
-        gf_list = []
-        logger.info('Ingesting lines from GFALL.')
-        for ion in self.gfall_ions:
+        """
+        Returns the same output than `AtomData._get_all_lines_data()`.
+        
+        The resulting DataFrame contains stacked transition lines for 
+        GFALL, Chianti (optional) and CMFGEN (optional).
 
-            try:
-                df = gf.lines.loc[ion]
+        """
 
-                # To match `line_id` field with the old API we keep
-                # track of how many GFALL lines we are skipping.
-                if ion in self.chianti_ions:
-                    df['line_id'] = range(start, len(df) + start)
-                    start += len(df)
-                    continue
+        logger.info('Ingesting transition lines.')
+        gf_lines = self.gfall_reader.lines
+        gf_lines['ds_id'] = 2
 
-                else:
-                    df['line_id'] = range(start, len(df) + start)
-                    start += len(df)
+        if self.chianti_reader is not None:
+            ch_lines = self.chianti_reader.lines
+            ch_lines['ds_id'] = 4
+        else:
+            ch_lines = pd.DataFrame(columns=gf_lines.columns)
 
-            except (KeyError, TypeError):
-                continue
+        if self.cmfgen_reader is not None:
+            cf_lines = self.cmfgen_reader.lines
+            cf_lines['ds_id'] = 5
+        else:
+            cf_lines = pd.DataFrame(columns=gf_lines.columns)
 
-            df = self.get_lvl_index2id(df, self.levels_all, ion)
-            df['source'] = 'gfall'
-            gf_list.append(df)
+        lines = pd.concat([gf_lines, ch_lines, cf_lines], sort=True)
+        lines = lines.reset_index()
+        lines = lines.rename(columns={'ion_charge': 'ion_number'})
+        lines['line_id'] = range(1, len(lines)+1)
 
-        ch_list = []
-        logger.info('Ingesting lines from Chianti.')
+        # Filter lines by priority
         for ion in self.chianti_ions:
+            mask = (lines['ds_id'] != 4) & (
+                        lines['atomic_number'] == ion[0]) & (
+                            lines['ion_number'] == ion[1])
+            lines = lines.drop(lines[mask].index)
+
+        for ion in self.cmfgen_ions:
+            mask = (lines['ds_id'] != 5) & (
+                        lines['atomic_number'] == ion[0]) & (
+                            lines['ion_number'] == ion[1])
+            lines = lines.drop(lines[mask].index)
+
+        lines = lines.set_index(['atomic_number', 'ion_number'])
```

Comment:
now it's `ion_number` - `ion_charge`?

Reply:
Always has been

![img](https://imgflip.com/s/meme/Always-Has-Been.png)